proposals/left-right-join-in-query-expressions.md

```diff
@@ -0,0 +1,427 @@
+	student => s.DepartmentID,
+	department => department.ID,
+	(student, department) => new { student.FirstName, student.LastName, Department = department.Name });
+
+// New left (outer) join operator: all students are returned, even those without any correlated department
+// Departments without a correlated student are not returned.
+var query = students.LeftJoin(
+	departments,
+	student => s.DepartmentID,
+	department => department.ID,
+	(student, department) => new { student.FirstName, student.LastName, Department = department?.Name });
+
+// New right (outer) join operator: all departments are returned, even those without any correlated student.
+// Students without a correlated department are not returned.
+var query = students.RightJoin(
+	departments,
+	student => s.DepartmentID,
+	department => department.ID,
+	(student, department) => new { student?.FirstName, student?.LastName, Department = department.Name });
+```
+
+Aside from allowing users to more easily express SQL LEFT/RIGHT JOIN when using a LINQ provider (such as EF Core), the new APIs also allow both easier and more efficient in-memory left/right joins, exactly in the way that inner joins are already supported.
+
+Along with the introduction of the operators into System.Linq, the C# query expression syntax can be extended with new `left` and `right` modifiers for the `join` clauses, which would translate to these new methods, allowing for simpler and more efficient code where C# query syntax is used.
+
+## Detailed design
+[design]: #detailed-design
+
+### Grammar
+
+Proposed grammar change ([§11.7.1](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1117-query-expressions)):
+
+```diff
+join_clause
+-    : 'join' type? identifier 'in' expression 'on' expression 'equals' expression
++    : ('left' | 'right')? 'join' type? identifier 'in' expression 'on' expression 'equals' expression
+    ;
+```
+
+The `join` clause is thus extended via new, optional `left` and `right` modifiers; the clause stays otherwise the same, with the modifiers only changing which LINQ method is translated to (see below):
+
+```c#
+from student in students
+left join department in departments on student.DepartmentID equals department.ID
+select new { student.Name, department?.Name }
+```
+
+### `join` clause specification
+
+This proposes removing the `join` clause from the [From, let, where, join and orderby clauses (§11.7.1)](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1117-query-expressions), and adding the following dedicated section for the `join` clause, including the new proposed modifiers.
+
+<details>
+<summary>Removal of `join` clause specs from [§11.7.1](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1117-query-expressions)</summary>
+
+A `join` clause immediately followed by a `select` clause
+
+```csharp
+from «x1» in «e1»  
+join «x2» in «e2» on «k1» equals «k2»  
+select «v»
+```
+
```

Comment:
remove lines 123 through 219 and say `... remainder of existing join specification up through ...`